File: apps/api/src/repositories/branch_repository.ts
import { PoolClient } from 'pg'
import { BranchDto, BranchCreateRequest } from '../dto/branch_dto'

export const branchRepository = {
  async create(client: PoolClient, data: BranchCreateRequest): Promise<BranchDto> {
    const status = data.status !== undefined ? data.status : true
    const sql = `
      INSERT INTO branches (id, business_id, name, status, created_at, updated_at)
      VALUES ($1, $2, $3, $4, now(), now())
      RETURNING id, business_id, name, status, created_at, updated_at
    `
    const result = await client.query(sql, [data.id, data.business_id, data.name, status])
    return result.rows[0] as BranchDto
  },

  async findAll(client: PoolClient, businessId: string): Promise<BranchDto[]> {
    const sql = `
      SELECT id, business_id, name, status, created_at, updated_at
      FROM branches
      WHERE business_id = $1
      ORDER BY created_at DESC
    `
    const result = await client.query(sql, [businessId])
    return result.rows as BranchDto[]
  },
  
  async findById(client: PoolClient, businessId: string, branchId: string): Promise<BranchDto | null> {
    const sql = `
      SELECT id, business_id, name, status, created_at, updated_at
      FROM branches
      WHERE business_id = $1 AND id = $2
    `
    const result = await client.query(sql, [businessId, branchId])
    return (result.rows[0] as BranchDto | undefined) ?? null
  }
}

File: apps/mobile/lib/core/sync/branch_repository.dart
import 'package:biz_erp_mobile/core/database/app_database.dart';
import 'package:biz_erp_mobile/core/sync/sync_models.dart';
import 'package:biz_erp_mobile/core/sync/sync_api_client.dart';
import 'package:drift/drift.dart';

class BranchRepository {
  BranchRepository(this._db, this._api);
  final AppDatabase _db;
  final SyncApiClient _api;

  Future<List<BranchDto>> fetchBranches(String businessId) async {
    final response = await _api.pullBranches(businessId: businessId);
    return response.branches;
  }

  Future<void> refreshAndCacheBranches(String businessId) async {
    final branches = await fetchBranches(businessId);
    if (branches.isEmpty) return;

    await _db.batch((b) {
      for (final branch in branches) {
        b.insert(
          _db.branchesLocal,
          BranchesLocalCompanion.insert(
            id: branch.id,
            businessId: branch.businessId,
            name: branch.name,
            status: branch.status,
            createdAt: branch.createdAt,
            updatedAt: branch.updatedAt,
            cachedAt: DateTime.now().millisecondsSinceEpoch,
          ),
          mode: InsertMode.insertOrReplace,
        );
      }
    });
  }

  Future<List<BranchDto>> getCachedBranches(String businessId) async {
    final rows = await (_db.select(_db.branchesLocal)
      ..where((t) => t.businessId.equals(businessId)))
      .get();
    return rows
        .map(
          (r) => BranchDto(
            id: r.id,
            businessId: r.businessId,
            name: r.name,
            status: r.status,
            createdAt: r.createdAt,
            updatedAt: r.updatedAt,
          ),
        )
        .toList();
  }

  Future<BranchDto?> getActiveBranch(String businessId) async {
    // Try to get cached branches first (for offline support)
    final cached = await getCachedBranches(businessId);
    if (cached.isNotEmpty) {
      // Return first active branch from cache
      for (final branch in cached) {
        if (branch.status) {
          return branch;
        }
      }
      return cached.first;
    }

    // If no cache, try online
    try {
      return await _getActiveBranchOnline(businessId);
    } catch (_) {
      // If online fails and no cache, return null
      return null;
    }
  }

  Future<BranchDto?> _getActiveBranchOnline(String businessId) async {
    final branches = await fetchBranches(businessId);
    if (branches.isEmpty) return null;

    // Cache the fetched branches
    await _db.batch((b) {
      for (final branch in branches) {
        b.insert(
          _db.branchesLocal,
          BranchesLocalCompanion.insert(
            id: branch.id,
            businessId: branch.businessId,
            name: branch.name,
            status: branch.status,
            createdAt: branch.createdAt,
            updatedAt: branch.updatedAt,
            cachedAt: DateTime.now().millisecondsSinceEpoch,
          ),
          mode: InsertMode.insertOrReplace,
        );
      }
    });

    for (final branch in branches) {
      if (branch.status) {
        return branch;
      }
    }
    return branches.first;
  }

  Future<void> setActiveBranch(String businessId, String branchId) async {
    await _db.into(_db.activeBranchLocal).insertOnConflictUpdate(
      ActiveBranchLocalCompanion.insert(
        businessId: businessId,
        branchId: branchId,
        updatedAt: DateTime.now().millisecondsSinceEpoch,
      ),
    );
  }

  Future<String?> getSelectedBranchId(String businessId) async {
    final row = await (_db.select(_db.activeBranchLocal)
      ..where((t) => t.businessId.equals(businessId)))
      .getSingleOrNull();
    return row?.branchId;
  }
}